### PR TITLE
new cut plane algo

### PR DESCRIPTION
**Pull request by arnaudon**
Reworked cut plane algorithm. 

- simplification of the implementation by dropping the CutPlane class
- different algorithm to estimate the probability of a cut plane (not just number of points in the cut, but number of leaves relative to expected numbers for the morphology)
- does not return json str, but the leaves and the qualities of the cut for each detected plane (the returned cut leaves can be on multiple planes)
- fixed 2 issues in the previous code (the bottom plane was not working due to a missing sign, and choice of best plane was not working due to incorrect value for coef_d before the max was taken)

### CODE REVIEW

File: neuror/cut_plane/detection.py [hunk 1]
```diff
@@ -205,48 +204,29 @@ def to_json(self):
...
-
-        Returns:
-            a numpy.histogram
-        '''
-        points = _get_points(self.morphology)
-        projected_points = self.project_on_directed_normal(points)
-        min_, max_ = np.min(projected_points, axis=0), np.max(projected_points, axis=0)
-        binning = np.arange(min_, max_ + self.bin_width, self.bin_width)
-        return np.histogram(projected_points, bins=binning)
+                'details': {'prob': self.prob, 'bin-width': self.bin_width},
```
Comment (adrien-berchet) [inline neuror/cut_plane/detection.py:209]: This looks like a breaking change so it should at least be added explicitly in the changelog.

Comment (arnaudon) [reply neuror/cut_plane/detection.py:209]: yes, this whole PR is very breaking change!

File: neuror/cut_plane/detection.py [hunk 2]
```diff
@@ -185,8 +181,11 @@ def cut_sections(self):
         leaves = np.array([leaf
                            for neurite in self.morphology.neurites
                            for leaf in iter_sections(neurite, iterator_type=Tree.ileaf)])
+        self.n_leaves = len(leaves)
         leaves_coord = [leaf.points[-1, COLS.XYZ] for leaf in leaves]
-        return leaves[self.distance(leaves_coord) < self.bin_width]
+        out = leaves[self.distance(leaves_coord) < self.bin_width]
```
Comment (arnaudon) [inline neuror/cut_plane/detection.py:186]: do we agree that `self.bin_with` is in fact half of it?

Comment (arnaudon) [reply neuror/cut_plane/detection.py:186]: the plane is on top of the morpho, so it is bin_width

**Comment by arnaudon** [comment]
I've added a quick test, @wizmer  do you think more tests are needed here?

**Comment by arnaudon** [comment]
@asanin-epfl , if you want to have a look here!

**Comment by arnaudon** [comment]
we need to fix sanitize now, I'll start the PR

**Comment by arnaudon** [comment]
@adrien-berchet , if you want to review that quickly as well! Also, if you can reproduce the test crash, that could be helpful, I cannot on bb5...

File: neuror/cut_plane/detection.py [hunk 3]
```diff
@@ -132,6 +133,9 @@ def find(cls, neuron, bin_width=3,
         Returns:
             A cut plane object
         """
+        warnings.warn("This method will be deprecated in favor or cut_leaves.find_cut_leaves"
```
Comment (wizmer) [inline neuror/cut_plane/detection.py:136]: I think you can simply drop all my shit

File: neuror/cut_plane/cut_leaves.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+from itertools import product
+import numpy as np
+from neurom.core.dataformat import COLS
+from neuror.cut_plane.planes import HalfSpace
+
+
+def _get_cut_leaves(plane, morphology, bin_width, percentile_threshold):
+    """Compute the cut leaves from a plane."""
+    # get the cut leaves
+    leaves = np.array([section for section in morphology.iter() if len(section.children) == 0])
```
Comment (wizmer) [inline neuror/cut_plane/cut_leaves.py:11]:
```suggestion
    leaves = np.array([section for section in morphology.iter() if not section.children])
```

**Review by wizmer — COMMENTED** [review]
You don't want to call the new method in main.py ?

File: neuror/cut_plane/detection.py (hunk 3 above)
Comment (arnaudon) [reply neuror/cut_plane/detection.py:136]: morphology-repair-workflow depends on it, no? Maybe better to test this new version a bit more, and make another PR to remove that + update morphokogy-repairr-workflow?

Comment (wizmer) [reply neuror/cut_plane/detection.py:136]: Ok, ok, yes I'm fine with this.

**Comment by arnaudon** [comment]
I had to manual hack a test because of this: https://github.com/BlueBrain/morph-tool/issues/59

**Comment by arnaudon** [comment]
> You don't want to call the new method in main.py ?

I added stuff in the init, is t a good idea?

**Comment by arnaudon** [comment]
> > You don't want to call the new method in main.py ?
> 
> I added stuff in the init, is t a good idea?

forget that, it is not

**Comment by arnaudon** [comment]
Could we merge that so I can move on to finalizing the other PR? thank you!

**Comment by asanin-epfl** [comment]
I'm reviewing currently

File: neuror/cut_plane/cut_leaves.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    planes = [
+        HalfSpace(int(axis == "X"), int(axis == "Y"), int(axis == "Z"), 0, upward=(side > 0))
+        for axis, side in product(searched_axes, searched_half_spaces)
+    ]
+
+    # set the plane coef_d as furthest morphology point
+    for plane, (axis, side) in zip(planes, product(searched_axes, searched_half_spaces)):
+        points = np.vstack([section.points for section in morph.iter()])
+        projected_points = plane.project_on_directed_normal(points)
+        plane.coefs[3] = -side * np.min(projected_points, axis=0)
```
[the remark this answers is not in the document]
Comment (arnaudon) [reply neuror/cut_plane/cut_leaves.py:72]: ah no,  the coefs are to represnet the plane, coef[0] * x +.. + coef[3], but the projected points is a list of points of shape (n, 3), so these axis have noting to do with each others.

File: neuror/cut_plane/cut_leaves.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    # create planes
+    searched_axes = [axis.upper() for axis in searched_axes]
+    planes = [
+        HalfSpace(int(axis == "X"), int(axis == "Y"), int(axis == "Z"), 0, upward=(side > 0))
+        for axis, side in product(searched_axes, searched_half_spaces)
+    ]
+
+    # set the plane coef_d as furthest morphology point
+    for plane, (axis, side) in zip(planes, product(searched_axes, searched_half_spaces)):
+        points = np.vstack([section.points for section in morph.iter()])
```
[the remark this answers is not in the document]
Comment (arnaudon) [reply neuror/cut_plane/cut_leaves.py:70]: ah yes, the original code of benoit was with neuron, but I prefer morphio, so I'll change that, good catch

File: neuror/cut_plane/cut_leaves.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    cuts = [_get_cut_leaves(plane, morph, bin_width, percentile_threshold) for plane in planes]
+
+    # return only leaves of planes with valid cut
+    leaves = [leave for leave, _ in cuts if leave is not None]
+    qualities = [
+        {"axis": axis, "side": side, "quality": np.around(quality, 3)}
+        for (_, quality), (axis, side) in zip(cuts, product(searched_axes, searched_half_spaces))
+        if quality is not None
+    ]
+    return np.vstack(leaves) if leaves else [], qualities
```
[the remark this answers is not in the document]
Comment (arnaudon) [reply neuror/cut_plane/cut_leaves.py:84]: really? I read
`( np.vstack(leaves) if leaves else []) , ( qualities)`
no?

File: neuror/cut_plane/cut_leaves.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    # compute the min cut leave given the percentile
+    projected_uncut_leaves = plane.project_on_directed_normal(leaves_coord[~cut_filter])
+    _min, _max = min(projected_uncut_leaves), max(projected_uncut_leaves)
+    bins = np.arange(_min, _max, bin_width)
+    _dig = np.digitize(projected_uncut_leaves, bins)
+    leaves_threshold = np.percentile(np.unique(_dig, return_counts=True)[1], percentile_threshold)
+
+    quality = len(cut_leaves) - leaves_threshold
+    if quality > 0:
```
[the remark this answers is not in the document]
Comment (arnaudon) [reply neuror/cut_plane/cut_leaves.py:24]: this is explained in the docstring, I used this term as a generic term: the higher, the more confident we are that there is a cut plane. Would you have a better name? In previous code, it was a probability computation, called `-logP`, not sure what is more explicit ;)

File: neuror/cut_plane/cut_leaves.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (asanin-epfl) [reply neuror/cut_plane/cut_leaves.py:82]: Please put brackets, besides `[]` and `np.vstack` still return different type.

File: neuror/cut_plane/cut_leaves.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (asanin-epfl) [reply neuror/cut_plane/cut_leaves.py:24]: It is not clear that `quality` is explained in the docstring of another function. At least make a comment that this var is explained in the docstring of `find_cut_leaves`.

File: neuror/cut_plane/cut_leaves.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (arnaudon) [reply neuror/cut_plane/cut_leaves.py:82]:
```
In [1]: def test():
   ...:     a=2
   ...:     return a if a==1 else [], 10
   ...: 

In [2]: test()
Out[2]: ([], 10)
```
but I agree, with brackets it's less confusin, and I'll add np.array()

File: neuror/cut_plane/cut_leaves.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (arnaudon) [reply neuror/cut_plane/cut_leaves.py:24]: yes, I'll improve that tomorrow!

File: neuror/cut_plane/cut_leaves.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (arnaudon) [reply neuror/cut_plane/cut_leaves.py]: I think it's more clear now, and it returns ndarray in both cases

File: neuror/cut_plane/cut_leaves.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (arnaudon) [reply neuror/cut_plane/cut_leaves.py:39]: I've made a proper docstring for this private function, so it's more clear. Indeed it was a bit obscure just reading that fucntion.

**Review by asanin-epfl — APPROVED** [review]
Thank you. Looks much easier to grasp now. Do you want to update Readme about this new API? Up to you. Approving.

**Comment by arnaudon** [comment]
So the plan is to deprecate the older version once this one is fully tested and used. I guess we can update readme once we make this one official?

**Comment by arnaudon** [comment]
I let you merge, I cannot on this repo, thanks for your review, very helpful!

**Pull request merged**